src/sparseml/pytorch/recipe_template/main.py

```diff
@@ -0,0 +1,351 @@
+        modifier_class = MovementPruningModifier
+        # movement pruning does not support global_sparsity
+        del pruning_args["global_sparsity"]
+
+    elif pruning_algo == "obs":
+        modifier_class = OBSPruningModifier
+        extra_recipe_variables.update(dict(num_grads=1024, fisher_block_size=50))
+        pruning_args.update(
+            dict(
+                num_grads="eval(num_grads)",
+                fisher_block_size="eval(fisher_block_size)",
+            )
+        )
+    elif pruning_algo == "constant":
+        modifier_class = ConstantPruningModifier
+        pruning_args = dict(start_epoch=0.0, end_epoch=10.0, params="__ALL_PRUNABLE__")
+    else:
+        raise ValueError(f"Unknown pruning_algo type {pruning_algo}")
+
+    pruning_modifier = ModifierYAMLBuilder(
+        modifier_class=modifier_class, **pruning_args
+    )
+    return [pruning_modifier], extra_recipe_variables
+
+
+def _get_prunable_params_for_module(model: Module) -> List[str]:
+    return [
+        f"{name}.weight"
+        for name, _ in get_prunable_layers(module=model)
+        if "input" not in name and "output" not in name
+    ]
+
+
+def _get_quantizable_params_for_module(model: Module) -> List[str]:
+    return [name for name, _ in get_quantizable_layers(module=model)]
+
+
+def _get_quantization_modifier_builders(
+    post_pruning: bool,
+    target: str,
+    model: Optional[Module] = None,
+) -> Tuple[List[ModifierYAMLBuilder], Dict[str, Any]]:
+    extra_recipe_variables = {}
+    freeze_graph_epoch = (
+        "eval(_num_pruning_epochs + num_qat_epochs - num_qat_finetuning_epochs)"
+        if post_pruning
+        else "eval(num_qat_epochs - num_qat_finetuning_epochs)"
+    )
+    quantization_modifier = ModifierYAMLBuilder(
+        modifier_class=QuantizationModifier,
+        start_epoch="eval(_num_pruning_epochs)" if post_pruning else 0.0,
+        submodules="eval(quantization_submodules)",
+        disable_quantization_observer_epoch=freeze_graph_epoch,
+        freeze_bn_stats_epoch=freeze_graph_epoch,
+        tensorrt=target == "tensorrt",
+        quantize_linear_activations=False,
+        quantize_conv_activations=False,
+        quantize_embedding_activations=False,
+        exclude_module_types=["LayerNorm", "Tanh"],
+    )
+
+    extra_recipe_variables["quantization_submodules"] = (
+        "null" if not model else _get_quantizable_params_for_module(model=model)
+    )
+    return [quantization_modifier], extra_recipe_variables
+
+
+def _validate_mask_type(
+    mask_type: str = "unstructured", target: Optional[str] = None
+) -> str:
+    target_to_mask_type = defaultdict(
+        lambda: "unstructured",
+        {"vnni": "block4", "tensorrt": "2:4"},
+    )
+
+    if mask_type != target_to_mask_type[target]:
+        warnings.warn(
+            f"The specified mask type {mask_type} and target {target} are "
+            f"incompatible, overriding mask_type tp {target_to_mask_type[target]}"
+        )
```

Comment:
Prefer exceptions instead of warnings which I am not convinced anyone ever sees 😁